### PR TITLE
Allow cloud-init to support creating btrfs partitions

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
###### Motivation for this change
`cloud-init` does not know where is `btrfs` programmes. It makes `btrfs` partition creation fails sometimes. This PR includes `btrfs-progs` into the `PATH` environment variable.

Request for comment: should we: 1. survey the all possible configurations and add in other missing runtime dependencies of `cloud-init`; 2. or, we add one option to allow users to supply additional runtime dependencies?

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
cc @madjar @phile314

**Review by phile314 — APPROVED** [review]
Looks good to me.

File: nixos/modules/services/system/cloud-init.nix [hunk 1]
```diff
@@ -3,7 +3,16 @@
 with lib;
 
 let cfg = config.services.cloud-init;
-    path = with pkgs; [ cloud-init nettools utillinux e2fsprogs shadow openssh iproute ];
+    path = with pkgs; [
+      btrfs-progs
```
Comment (Mic92) [inline nixos/modules/services/system/cloud-init.nix]: `btrfs-progs` adds like `3.5mb`. I think we can add an option called `btrfs.enable` (default: false) and an option `ext4.enable` (default: true).

Comment (dingxiangfei2009) [reply nixos/modules/services/system/cloud-init.nix]: @Mic92 Sure. Sounds like a good idea to me.

I will re-write this as a module in the coming commits of this PR.

**Comment by Mic92** [comment]
@GrahamcOfBorg test cloud-init

**Comment by Mic92** [comment]
Looks like a bug in ofborg

**Comment by grahamc** [comment]
OfBorg hasn't changed, must be a change in Nixpkgs.

**Comment by Ekleog** [comment]
Bisecting confirms it is, kind of unsurprisingly given it's where the stack trace appeared, 83b27f60ceff23967e477c90bef8e78cc96d50a2.

What is *really* weird is the following:
```
$ nix-build -I . --option restrict-eval true nixos/release.nix -A tests.env --show-trace
... fails with the error message
$ nix-build -I . --option restrict-eval true nixos/tests/env.nix --show-trace
... succeeds
```

Which would appear to imply that the problem comes from the way `release.nix` imports the test, ie. with `import`ing from its `nixpkgs` argument. *But* hydra also builds the `env` test as taken from `release.nix` with `restrict-eval`, and the change appears not to have hindered the latest evals / rebuilds.

So I don't really understand how & why the problem would arise from https://github.com/NixOS/nixpkgs/blob/master/nixos/release.nix#L21-L27 except on hydra. Especially given the fact that the import is the same than the one at https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/make-test.nix#L3, unless the passed `nixpkgs` is some kind of trick (I've tried adding the `config = {};` locally just in case without more success)

Now, let's see if the problem arises with https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L76-L80 too… it doesn't locally. To add to the confusion.

@GrahamcOfBorg build nixosTests.cloud-init

(cc @roberth given our ongoing discussion on https://github.com/NixOS/nixpkgs/issues/50301, in case you have an idea about where this could be going wrong)

TL;DR:
 * This appeared in 83b27f60ceff23967e477c90bef8e78cc96d50a2
 * https://github.com/NixOS/nixpkgs/blob/master/nixos/release.nix#L21-L27 fails to eval on `restrict-eval`… except on hydra
 * https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L76-L80 doesn't fail

**Comment by Ekleog** [comment]
Correction: hydra builds do fail due to this. Hydra just doesn't report failed evaluations as failures.

**Comment by Ekleog** [comment]
@Mic92 @grahamc @roberth The issue with ofborg building tests (and hopefully hydra too) should be fixed with https://github.com/NixOS/nixpkgs/pull/50342. It appears, if I understand correctly what I've been doing, to have stemmed from an IFD that had stayed unnoticed until now because we never actually used it for the tests: the `cleanSource` at the top of `release.nix`

**Comment by grahamc** [comment]
Please have @edolstra review

File: nixos/modules/services/system/cloud-init.nix [hunk 2]
```diff
@@ -29,6 +36,22 @@ in
         '';
       };
 
+      btrfs.enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Allow the cloud-init service to operate `btrfs` filesystem.
```
Comment (edolstra) [inline nixos/modules/services/system/cloud-init.nix]: BTW, what does it mean to "operate" a filesystem?
